source/load-balancers/tests/transactions.yml

```diff
@@ -0,0 +1,447 @@
+          session: *session0
+      - name: assertNumberConnectionsCheckedOut
+        object: testRunner
+        arguments:
+          client: *client0
+          connections: 1
+      - name: commitTransaction
+        object: *session0
+      - name: assertNumberConnectionsCheckedOut
+        object: testRunner
+        arguments:
+          client: *client0
+          connections: 0
+    expectEvents:
+      - client: *client0
+        events:
+          - commandStartedEvent:
+              command:
+                insert: *collection0Name
+                documents: [*insertDocument]
+              commandName: insert
+          - commandStartedEvent:
+              commandName: commitTransaction
+          - connectionReadyEvent: {}
+          - connectionCheckedOutEvent: {}
+          - connectionCheckedInEvent: {}
+
+  - description: pinned connections are returned on abort
+    operations:
+      - name: startTransaction
+        object: *session0
+      - name: insertOne
+        object: *collection0
+        arguments:
+          document: *insertDocument
+          session: *session0
+      - name: assertNumberConnectionsCheckedOut
+        object: testRunner
+        arguments:
+          client: *client0
+          connections: 1
+      - name: abortTransaction
+        object: *session0
+      - name: assertNumberConnectionsCheckedOut
+        object: testRunner
+        arguments:
+          client: *client0
+          connections: 0
+    expectEvents:
+      - client: *client0
+        events:
+          - commandStartedEvent:
+              command:
+                insert: *collection0Name
+                documents: [*insertDocument]
+              commandName: insert
+          - commandStartedEvent:
+              commandName: abortTransaction
+          - connectionReadyEvent: {}
+          - connectionCheckedOutEvent: {}
+          - connectionCheckedInEvent: {}
+
+  - description: commitTransaction can be retried after a network error
+    operations:
+      - name: failPoint
+        object: testRunner
+        arguments:
+          client: *client0
+          failPoint:
+            configureFailPoint: failCommand
+            mode: { times: 1 }
+            data:
+              failCommands: [commitTransaction]
+              closeConnection: true
+      - name: startTransaction
+        object: *session0
+      - name: insertOne
+        object: *collection0
+        arguments:
+          document: *insertDocument
+          session: *session0
+      - name: assertNumberConnectionsCheckedOut
+        object: testRunner
+        arguments:
+          client: *client0
+          connections: 1
+      - name: commitTransaction
+        object: *session0
+        ignoreResultAndError: true
+      - name: assertNumberConnectionsCheckedOut
+        object: testRunner
+        arguments:
+          client: *client0
+          connections: 0
+    expectEvents:
+      - client: *client0
+        events:
```

Comment:
I don't expect test runners to store multiple event types in a single array and correlate the timing. Instead, I expect APM events and CMAP events to be stored in separate arrays and assertions to be made separately. The reasoning behind writing out the expectations separately was just to make it easier to read. I can clarify this in the actual test format spec if you think that would be helpful.

Reply:
This definitely needs to be documented.

Given that [`expectEvents`](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#test) is an array of one or more [`expectedEventsForClient`](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#expectedeventsforclient) objects, I think it'd be preferable to segregate these events into their own `expectedEventsForClient` objects.

Then, we can maintain language advising tests not to mix event types within `expectedEventsForClient` and suggest the approach above. I'm OK with removing the advice against mixing event types in `observeEvents`.

One last idea is that we may want to enforce the type of events within `expectedEventsForClient`. We could do that by introducing a new optional field (e.g. `type`), which is an enum ("command" and "cmap" for now, potentially "sdam" down the line) and defaults to "command" if omitted for BC. This would mean that drivers could never mix events from different specs in the same list of expectations, but I think that's reasonable since there is likely too much variation to ever standardize on that.

Reply:
I pushed a change to add a new `eventType` field and specify which events can be present based on the type.